src/coprocessor/dag/expr/arithmetic.rs

```diff
@@ -0,0 +1,390 @@
+        let lhs = try!(self.children[0].eval_int(ctx, row));
+        let rhs = try!(self.children[1].eval_int(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let res = (l as u64).checked_add(r as u64).ok_or(Error::Overflow);
+            res.map(|u| u as i64)
+        })
+    }
+
+    pub fn minus_real(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<f64>> {
+        let lhs = try!(self.children[0].eval_real(ctx, row));
+        let rhs = try!(self.children[1].eval_real(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let res = l - r;
+            if !res.is_finite() {
+                return Err(Error::Overflow);
+            }
+            Ok(res)
+        })
+    }
+
+    pub fn minus_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try!(self.children[1].eval_decimal(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let result: Result<Decimal> = (l.as_ref() - r.as_ref()).into();
+            result.map(Cow::Owned)
+        })
+    }
+
+    pub fn minus_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_int(ctx, row));
+        let rhs = try!(self.children[1].eval_int(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| l.checked_sub(r).ok_or(Error::Overflow))
+    }
+
+    pub fn minus_uint(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_int(ctx, row));
+        let rhs = try!(self.children[1].eval_int(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let res = (l as u64).checked_sub(r as u64).ok_or(Error::Overflow);
+            res.map(|u| u as i64)
+        })
+    }
+
+    pub fn multiply_real(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<f64>> {
+        let lhs = try!(self.children[0].eval_real(ctx, row));
+        let rhs = try!(self.children[1].eval_real(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let res = l * r;
+            if !res.is_finite() {
+                return Err(Error::Overflow);
+            }
+            Ok(res)
+        })
+    }
+
+    pub fn multiply_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try!(self.children[1].eval_decimal(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let result: Result<Decimal> = (l.as_ref() * r.as_ref()).into();
+            result.map(Cow::Owned)
+        })
+    }
+
+    pub fn multiply_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_int(ctx, row));
```

Comment:
you mean  `1u64 * -1i64`?  TiDB will deal with this case by converting both of them to `u64` or `i64`. So We won't get `u64 * i64` or `i64 * u64` in any signatures. We only have `MultiplyInt` and `MultiplyIntUnsigned`, which deal with `i64` and `u64` respectively.